### PR TITLE
QUIC deferred logging: add retransmission rate

### PR DESCRIPTION
**Pull request by pksohn**
Commit Message: [QUIC only] Log retransmitted packets and bytes.
Additional Description: Adds to the deferred logging implementation from #23648 and implements the existing OnPacketRetransmitted function.

Risk Level: Low
Testing: Integration test
Docs Changes: N/A
Release Notes: added

### CODE REVIEW

File: changelogs/current.yaml [hunk 1]
```diff
@@ -160,6 +160,9 @@ new_features:
 - area: access_log
   change: |
     added :ref:`CEL <envoy_v3_api_msg_extensions.formatter.cel.v3.Cel>` access log formatter to print CEL expression.
+- area: access_log
+  change: |
+    (QUIC only) Added support for %BYTES_RETRANSMITTED% and %PACKETS_RETRANSMITTED%.
```
Comment (alyssawilk) [inline changelogs/current.yaml:165]: can you update logging docs to cover these as well?

File: envoy/stream_info/stream_info.h [hunk 2]
```diff
@@ -580,6 +580,26 @@ class StreamInfo {
    */
   virtual uint64_t bytesReceived() const PURE;
 
+  /**
+   * @param bytes_retransmitted denotes number of bytes to add to total retransmitted bytes.
+   */
+  virtual void addBytesRetransmitted(uint64_t bytes_retransmitted) PURE;
+
+  /**
+   * @return the number of body bytes retransmitted by the stream.
```
Comment (alyssawilk) [inline envoy/stream_info/stream_info.h:589]: nothing else in here calls out it's body bytes.  do we not track header bytes retransmitted? if not thoughts on naming?

File: source/common/stream_info/stream_info_impl.h [hunk 3]
```diff
@@ -185,6 +185,18 @@ struct StreamInfoImpl : public StreamInfo {
 
   uint64_t bytesReceived() const override { return bytes_received_; }
 
+  void addBytesRetransmitted(uint64_t bytes_retransmitted) override {
```
Comment (alyssawilk) [inline source/common/stream_info/stream_info_impl.h:188]: I'm tentatively Ok with this but do we plan on any other HTTP/3 specific stats?  If so I'd suggest we stick them in an optional struct and lazy initialize the memory.

File: source/common/stream_info/stream_info_impl.h [hunk 4]
```diff
@@ -435,6 +449,8 @@ struct StreamInfoImpl : public StreamInfo {
 
   std::shared_ptr<UpstreamInfo> upstream_info_;
   uint64_t bytes_received_{};
+  uint64_t bytes_retransmitted_{};
```
Comment (alyssawilk) [inline source/common/stream_info/stream_info_impl.h:452]: bah.  I was going to ask for downstream-specific naming but looks like that ship has sailed =P

**Review by alyssawilk — COMMENTED** [review]
This looks great overall!  nits follow
/wait

File: source/common/stream_info/stream_info_impl.h (hunk 3 above)
Comment (pksohn) [reply source/common/stream_info/stream_info_impl.h:188]: No further H/3 stats planned; this completes the implementation of the [QuicAckListenerInterface](https://github.com/google/quiche/blob/main/quiche/quic/core/quic_ack_listener_interface.h). Furthermore I hope to generalize these to H/1 and H/2.

File: envoy/stream_info/stream_info.h (hunk 2 above)
Comment (alyssawilk) [reply envoy/stream_info/stream_info.h:589]: If you're planning to implement this for HTTP/1 and 2 I assume they're going to track all bytes, in which case I'd be inclined to keep the naming the same and comment that for HTTP/3 it only tracks body bytes?
actually what's the plan for H2 given things are intermingled on the same TCP connection?
cc @RyanTheOptimist

File: changelogs/current.yaml (hunk 1 above)
Comment (pksohn) [reply changelogs/current.yaml:165]: Done!

**Comment by alyssawilk** [comment]
@zuercher ping?
/retest

**Review by zuercher — APPROVED** [review]
lgtm

**Pull request merged**